internal/core/thirdparty/tantivy/tantivy-binding/src/analyzer/dict/lindera/fetch.rs

```diff
@@ -265,19 +251,64 @@ pub async fn fetch(
         empty_directory(&output_dir)?;
 
         // Rename tmp_path to output_dir
-        rename(tmp_path, &output_dir)?;
+        fs::rename(tmp_path, &output_dir)?;
     }
-
-    let _ = std::fs::remove_dir_all(&input_dir);
-
     info!(
-        "finish fetch lindera dictionary name: {} duration: {} ms\n",
+        "finish donwload and build lindera dictionary. name: {} duration: {} ms\n",
         params.file_name.as_str(),
         start.elapsed().as_millis()
     );
     Ok(())
 }
 
+/// Fetch the necessary assets and then build the dictionary using `builder`
+pub async fn fetch(
+    params: &FetchParams,
+    builder: impl DictionaryBuilder,
+) -> Result<(), Box<dyn Error>> {
+    let build_dir = PathBuf::from(params.lindera_dir.as_str());
+    std::fs::create_dir_all(&build_dir)?;
+
+    let input_dir = build_dir.join(params.input_dir.as_str());
+    let output_dir = build_dir.join(params.output_dir.as_str());
+    let lock_path = build_dir.join(format!("lindera-{}.lock", params.file_name.as_str()));
+
+    // Skip create fs lock if already in cache
+    if output_dir.is_dir() {
+        return Ok(());
+    }
+
+    let flock = fs::File::create(&lock_path)?;
+    flock.lock_exclusive()?;
```

Comment:
try_lock_exclusive is better here? And it seems that this lock is only blocking among process not threads. Is this what you want?